compiler/front_end/parser.py

```diff
@@ -12,120 +12,42 @@
-    for error_example in error_examples[1:]:
-        message_and_examples = error_example.split("\n" + "-" * 80 + "\n")
-        if len(message_and_examples) != 2:
-            raise ParserGenerationError(
-                "Expected one error message and one example section in:\n"
-                + error_example
-            )
-        message, example_text = message_and_examples
-        examples = example_text.split("\n---\n")
-        for example in examples:
-            # TODO(bolms): feed a line number into tokenize, so that tokenization
-            # failures refer to the correct line within error_example_text.
-            tokens, errors = tokenizer.tokenize(example, "")
-            if errors:
-                raise ParserGenerationError(str(errors))
-
-            for i in range(len(tokens)):
-                if tokens[i].symbol == "BadWord" and tokens[i].text == "$ANY":
-                    tokens[i] = lr1.ANY_TOKEN
-
-            error_token = None
-            for i in range(len(tokens)):
-                if tokens[i].symbol == "BadWord" and tokens[i].text == "$ERR":
-                    error_token = tokens[i + 1]
-                    del tokens[i]
-                    break
-            else:
-                raise ParserGenerationError(
-                    "No error token marker '$ERR' in:\n" + error_example
-                )
-
-            result.append((tokens, error_token, message.strip(), example))
-    return result
-
-
-def generate_parser(start_symbol, productions, error_examples):
-    """Generates a parser from grammar, and applies error_examples.
-
-    Arguments:
-        start_symbol: the start symbol of the grammar (a string)
-        productions: a list of parser_types.Production in the grammar
-        error_examples: A list of (source tokens, error message, source text)
-            tuples.
-
-    Returns:
-        A parser.
-
-    Raises:
-        ParserGenerationError: There is a problem generating the parser.
-    """
-    parser = lr1.Grammar(start_symbol, productions).parser()
-    if parser.conflicts:
-        raise ParserGenerationError("\n".join([str(c) for c in parser.conflicts]))
-    for example in error_examples:
-        mark_result = parser.mark_error(example[0], example[1], example[2])
-        if mark_result:
-            raise ParserGenerationError(
-                "error marking example: {}\nExample:\n{}".format(
-                    mark_result, example[3]
-                )
-            )
-    return parser
-
-
 @simple_memoizer.memoize
-def _load_module_parser():
-    error_examples = parse_error_examples(
-        resources.load("compiler.front_end", "error_examples")
-    )
-    return generate_parser(
-        module_ir.START_SYMBOL, module_ir.PRODUCTIONS, error_examples
-    )
-
-
-@simple_memoizer.memoize
-def _load_expression_parser():
-    return generate_parser(module_ir.EXPRESSION_START_SYMBOL, module_ir.PRODUCTIONS, [])
+def _load_parsers():
```

Comment:
med: can we add a test for some of this?

Reply:
I am not entirely sure what you mean.

If you're asking for tests for the case where `module_ir` no longer matches the cached parser... in theory, yes, but I don't see a particularly easy or clean way that really gets good coverage.  (On the other hand, if it is broken, a developer who makes a change to `module_ir` should notice fairly quickly that pretty much every test in the suite is broken, and be able to debug that way.)

I think the normal case, where the cached parser does match, is covered by existing tests.

Reply:
Resolved per external conversation: the determinism of the generated parser is (implicitly) covered by `cached_parser_is_up_to_date_test`.